### PR TITLE
Remove pkgs aliases

### PR DESCRIPTION
**Pull request by nbp**
This branch aims at keeping the same hashes, but only removing aliases reported by the static analysis added in #10851 .

This branch aims at solving one of the problem which should be fixed to make the #10851 hypothesis hold, which is that all packages are defined in the first "generation" of Nixpkgs evaluation.

Thus transforming the following attribute accesses:

```
   +-----------+
   |           |
   |  nixpkgs  |
   |           |        +-----------+
   |   gen 3   |        |           |
   |           |        |  nixpkgs  |
   +----^------+        |           |        +-----------+
        |               |   gen 2   |        |           |   nix-build -A foo
        +---------------+           |        |  nixpkgs  |                 +
  foo_4 = callPackage…  +----^------+        |           |                 |
                             |               |   gen 1   <-----------------+
                             +---------------+           |
                         foo = foo_4;        +-----------+
```

Into something like:

```
                        +-----------+
                        |           |
                        |  nixpkgs  |
                        |           |
                        |   gen 2   |
                        |           |
                        +----^------+
                             |
                             +------------------+
                       foo_4 = callPackage…     |
                                             +-----------+
                                             |           |   nix-build -A foo
                                             |  nixpkgs  |                 +
                             +--------------->           |                 |
                             |               |   gen 1   <-----------------+
                             +---------------+           |
                         foo = self.foo_4;   +-----------+
```

These changes are likely to bit-rot over time, thus this pull request is a meta issue / staging area for patches which are pending for review.

cc @edolstra

update: rephrase, to make statements clearer

### CODE REVIEW

**Comment by domenkozar** [comment]
Can we come up with a test that will check this is true for future commits?

**Comment by nbp** [comment]
@domenkozar Yes, my initial plan was to use the static analysis to ensure that the number issues does not grow.  But we can only do that as soon as #10851 is landed, and as soon as the number of issues is low (no bad idioms remain).

**Comment by aneeshusa** [comment]
In my nixpkgs config I make heavy use of `packageOverrides`, using `override` and `overrideDerivation`, as well as redefining some aliases, e.g. I set `openssl = pkgs.libressl_2_3`. Does this PR change the best way to set aliases with custom overrides? Is there a way I can run the static analysis myself?

`packageOverrides` doesn't seem to make a `self` available in scope, but `pkgs.overridePackages` does. Should I switch which function I'm using?

Current docs: https://nixos.org/nixpkgs/manual/#sec-modify-via-packageOverrides

**Comment by nbp** [comment]
@aneeshusa any package defined in `packageOverrides` is using the newly defined packages (`pkgs` corresponds to `super`) Thus, setting your own packages there should be fine, as long as you are not making aliases in your own configuration file.

However, if Nixpkgs defines `openssl = self.openssl_1_0_2;` and you set the attribute `openssl_1_0_2` in `packageOverrides`, then it will no longer be taken into account by packages which are using `openssl`, you will have to add the alias in `packageOverrides` too.

The static analysis already use `packageOverrides` as inputs, so yes this will work.  On the other hand, if you run it today, you will see ~21k issues, which are Nixpkgs issues.

No, you should not switch to `overridePackages`, which is much more costly.  At the moment, Nixpkgs's `self` is a *weaker* binding equivalent to `rec`.

**Comment by nbp** [comment]
@vcunat You can run the static analysis by checking out the `security-updates` branch from my repository.  Then you can run the static analysis with the following command:

```
$ nix-instantiate ./pkgs/test/security/static-analaysis.nix --eval-only --show-trace
```

Take care, this command take quite some time, and it is quite verbose as well, so you might want to redirect the output into a file.

This branch aims at fixing issues which are like `alias-original: generation 1`, which are easily fixable in `all-packages.nix`. They are a few others which need more thinking, but leave them aside for the moment.

Then, I guess you can make a pull request to the `remove-pkgs-aliases` branch [2] in my repository, and it should appear here as soon as I merge it.

[1] https://github.com/nbp/nixpkgs-2/tree/security-updates
[1] https://github.com/nbp/nixpkgs-2/tree/remove-pkgs-aliases

**Comment by jagajaga** [comment]
cc me

**Comment by C0DEHERO** [comment]
So if I understand this correctly, you just prepend "self." to any self references in a package?

**Comment by nbp** [comment]
> So if I understand this correctly, you just prepend "self." to any self references in a package?

Yes, for any top-level self reference and as long as it does not change the hashes, which is mostly the case, except for packages which are overwitten by the `stdenvOverride`.  I yet have to find what to do for these, but I guess we should make dedicated pull request for these cases.

File: pkgs/top-level/all-packages.nix [hunk 1]
```diff
@@ -10060,11 +10060,8 @@ in
   microcodeIntel = callPackage ../os-specific/linux/microcode/intel.nix { };
 
   apparmor = callPackage ../os-specific/linux/apparmor { swig = swig2; };
```
Comment (edolstra) [inline pkgs/top-level/all-packages.nix]: This should use `callPackages`, i.e.
```
  inherit (callPackages ../os-specific/linux/apparmor { ... }) libapparmor ...;
```

**Comment by edolstra** [comment]
Isn't all that's required to prevent self-references in the future to remove the `with pkgs` from `all-packages.nix`?

**Comment by nbp** [comment]
@edolstra 
> Isn't all that's required to prevent self-references in the future to remove the `with pkgs` from `all-packages.nix`?

Technically yes, but `with pkgs` might also be convenient for non automated arguments.

For example, the following code:

```nix
  foo_4 = callPackage ../path/to/foo { swig = swig2; };
  foo = foo_4;
```

should be changed into:

```nix
  foo_4 = callPackage ../path/to/foo { swig = pkgs.swig2; };
  foo = self.foo_4;
```

but not into:

```nix
  foo_4 = callPackage ../path/to/foo { swig = self.swig2; };
  foo = self.foo_4;
```

At least, if we are building dynamically against `swig2`, and not statically.

So removing `with pkgs` would imply that we have to explicit all the non-default arguments, to be taken from `pkgs`.

**Comment by abbradar** [comment]
Those changes seem both trivial (and shouldn't change hashes if I understand correctly) and affecting fast-moving target (`all-packages.nix`). I suppose we should merge this ASAP and push further changes like this directly to `master` to ease burden on @nbp and others who need to keep the branch mergeable -- what do you think?

**Pull request merged**